frame/support/procedural/src/pallet/expand/documentation.rs

```diff
@@ -0,0 +1,211 @@
+use syn::{
+	parenthesized,
+	parse::{self, Parse, ParseStream},
+	spanned::Spanned,
+	Attribute, Lit, Token,
+};
+
+const DOC: &'static str = "doc";
+const INCLUDE_STR: &'static str = "include_str";
+const PALLET_DOC: &'static str = "pallet_doc";
+
+/// Get the documentation file path from the `pallet_doc` attribute.
+///
+/// Supported format:
+/// `#[pallet_doc(PATH)]`: The path of the file from which the documentation is loaded
+fn parse_pallet_doc_value(attr: &Attribute) -> syn::Result<DocMetaValue> {
+	let span = attr.span();
+
+	let meta = attr.parse_meta()?;
+	let syn::Meta::List(metalist) = meta else {
+                let msg =
+                        "The `pallet_doc` attribute must receive arguments as a list. Supported format: `pallet_doc(PATH)`";
+                return Err(syn::Error::new(span, msg))
+        };
+
+	let paths: Vec<_> = metalist
+                .nested
+                .into_iter()
+                .map(|nested| {
+                        let syn::NestedMeta::Lit(lit) = nested else {
+                                let msg = "The `pallet_doc` received an unsupported argument. Supported format: `pallet_doc(PATH)`";
+                        return Err(syn::Error::new(span, msg))
+                };
+
+                        Ok(lit)
+                })
+                .collect::<syn::Result<_>>()?;
+
+	if paths.len() != 1 {
+		let msg = "The `pallet_doc` attribute must receive only one argument. Supported format: `pallet_doc(PATH)`";
+		return Err(syn::Error::new(span, msg))
+	}
+
+	Ok(DocMetaValue::Path(paths[0].clone()))
+}
+
+/// Get the value from the `doc` comment attribute:
+///
+/// Supported formats:
+/// - `#[doc = "A doc string"]`: Documentation as a string literal
+/// - `#[doc = include_str!(PATH)]`: Documentation obtained from a path
+fn parse_doc_value(attr: &Attribute) -> Option<DocMetaValue> {
+	let Some(ident) = attr.path.get_ident() else {
+                return None
+        };
+	if ident != DOC {
+		return None
+	}
+
+	let parser = |input: ParseStream| DocMetaValue::parse(input);
+	parse::Parser::parse2(parser, attr.tokens.clone()).ok()
+}
+
+/// Supported documentation tokens.
+#[derive(Debug)]
+enum DocMetaValue {
+	/// Documentation with string literals.
+	///
+	/// `#[doc = "Lit"]`
+	Lit(Lit),
+	/// Documentation with `include_str!` macro.
+	///
+	/// The string literal represents the file `PATH`.
+	///
+	/// `#[doc = include_str!(PATH)]`
+	Path(Lit),
+}
+
+impl Parse for DocMetaValue {
+	fn parse(input: ParseStream) -> syn::Result<Self> {
+		let _token: Token![=] = input.parse()?;
+
+		if input.peek(Lit) {
+			return input.parse().map(DocMetaValue::Lit)
+		}
+
+		let ident: Ident = input.parse()?;
+		if ident != INCLUDE_STR {
+			return Err(input.error("expected include_str ident"))
+		}
+		let _token: Token![!] = input.parse()?;
+
+		// We must have a path literal inside `(...)`
+		let content;
+		parenthesized!(content in input);
+		content.parse().map(DocMetaValue::Path)
+	}
```

Comment:
btw you could use `derive_syn_parse` to do this very compactly. It has facilities for parsing enums and peeking, not that there is anything wrong with deriving `Parse` manually

Reply:
That makes sense! I've used `derive_syn_parse` for `doc = ` parsing only, for the `pallet_doc` I feel the errors are more explicit with the manual parsing, thanks for the review!